### PR TITLE
PIM-4416: Add comparators for product fields

### PR DESCRIPTION
**Pull request by skeleton**
| Q                    | A
| -------------------- | ---
| Bug fix?             | N
| New feature?         | N
| BC breaks?           | N
| CI currently passes? | ?
| Tests pass?          | Y
| Scenarios pass?      | Y
| Checkstyle issues?*  | N
| PMD issues?**        | N
| Changelog updated?   | N
| Fixed tickets        | PIM-4416
| DB schema updated?   | N
| Migration script?    | N
| Doc PR               | N

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/Resources/config/comparators.yml [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
-    pim_catalog.comparator.boolean.class:  Pim\Component\Catalog\Comparator\BooleanComparator
+    pim_catalog.comparator.registry.class:       Pim\Component\Catalog\Comparator\ComparatorRegistry
+    pim_catalog.comparator.media.class:          Pim\Component\Catalog\Comparator\MediaComparator
+    pim_catalog.comparator.prices.class:         Pim\Component\Catalog\Comparator\PricesComparator
+    pim_catalog.comparator.metric.class:         Pim\Component\Catalog\Comparator\MetricComparator
+    pim_catalog.comparator.option.class:         Pim\Component\Catalog\Comparator\OptionComparator
+    pim_catalog.comparator.options.class:        Pim\Component\Catalog\Comparator\OptionsComparator
+    pim_catalog.comparator.scalar.class:         Pim\Component\Catalog\Comparator\ScalarComparator
+    pim_catalog.comparator.boolean.class:        Pim\Component\Catalog\Comparator\BooleanComparator
+    pim_catalog.comparator.field.boolean.class:  Pim\Component\Catalog\Comparator\Field\BooleanComparator
```
Comment (nidup) [inline src/Pim/Bundle/CatalogBundle/Resources/config/comparators.yml]: weird to have pim_catalog.comparator.boolean.class and pim_catalog.comparator.field.boolean.class looks not consistent, should we prefix also the attribute/value comparator

File: src/Pim/Component/Catalog/Comparator/BooleanComparator.php [hunk 2]
```diff
@@ -24,7 +24,7 @@ public function supports($type)
     /**
      * {@inheritdoc}
      */
-    public function compare(array $data, array $originals)
+    public function compare($data, $originals)
```
Comment (nidup) [inline src/Pim/Component/Catalog/Comparator/BooleanComparator.php]: why no typehint?

File: src/Pim/Bundle/CatalogBundle/Resources/config/comparators.yml (hunk 1 above)
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Resources/config/comparators.yml]: OK, i did not rename them for BC, but i can do it

File: src/Pim/Component/Catalog/Comparator/BooleanComparator.php (hunk 2 above)
Comment (skeleton) [inline src/Pim/Component/Catalog/Comparator/BooleanComparator.php]: New comparators are not array, it's a string.

File: src/Pim/Component/Catalog/Comparator/ComparatorInterface.php [hunk 3]
```diff
@@ -24,10 +24,10 @@ public function supports($data);
      * Get the changes between a normalized product value instance and the updated data
      * If no changes detected, then the method returns null
      *
-     * @param array $data
-     * @param array $originals
+     * @param mixed $data
```
Comment (solivier) [inline src/Pim/Component/Catalog/Comparator/ComparatorInterface.php]: it's weird we now accept mixed data. IMO it's a bad idea to change a strong typed method signature to accept all data.

**Pull request merged**